relays/ethereum/src/message_race_delivery.rs

```diff
@@ -135,34 +136,53 @@ where
 	}
 }
 
-/// Message delivery strategy.
-struct MessageDeliveryStrategy<P: MessageLane> {
+/// Messages delivery strategy.
+type MessageDeliveryStrategy<P> = DeliveryStrategy<
+	<P as MessageLane>::SourceHeaderNumber,
+	<P as MessageLane>::SourceHeaderHash,
+	<P as MessageLane>::TargetHeaderNumber,
+	<P as MessageLane>::TargetHeaderHash,
+	<P as MessageLane>::MessageNonce,
+	<P as MessageLane>::MessagesProof,
+>;
+
+/// Nonces delivery strategy.
+#[derive(Debug)]
+pub struct DeliveryStrategy<SHN, SHH, THN, THH, Nonce, Proof> {
 	/// All queued nonces.
-	source_queue: VecDeque<(SourceHeaderIdOf<P>, P::MessageNonce)>,
+	source_queue: VecDeque<(HeaderId<SHH, SHN>, Nonce)>,
 	/// Best nonce known to target node.
-	target_nonce: P::MessageNonce,
+	target_nonce: Nonce,
+	/// Max nonces to relay in single transaction.
+	max_nonces_to_relay_in_single_tx: Nonce,
 	/// Unused generic types dump.
-	_phantom: PhantomData<P>,
+	_phantom: PhantomData<(THN, THH, Proof)>,
 }
 
-impl<P: MessageLane> Default for MessageDeliveryStrategy<P> {
-	fn default() -> Self {
-		MessageDeliveryStrategy {
+impl<SHN, SHH, THN, THH, Nonce: Default, Proof> DeliveryStrategy<SHN, SHH, THN, THH, Nonce, Proof> {
+	/// Create new delivery strategy.
+	pub fn new(max_nonces_to_relay_in_single_tx: Nonce) -> Self {
+		DeliveryStrategy {
 			source_queue: VecDeque::new(),
 			target_nonce: Default::default(),
+			max_nonces_to_relay_in_single_tx,
 			_phantom: Default::default(),
 		}
 	}
 }
 
-impl<P: MessageLane> RaceStrategy<SourceHeaderIdOf<P>, TargetHeaderIdOf<P>, P::MessageNonce, P::MessagesProof>
-	for MessageDeliveryStrategy<P>
+impl<SHN, SHH, THN, THH, Nonce, Proof> RaceStrategy<HeaderId<SHH, SHN>, HeaderId<THH, THN>, Nonce, Proof>
```

Comment:
Yeah, so I know I was the one that suggested not using full names for generics, but in this case this makes the code more confusing since now you need to think about what these are acronyms for (it's not clear from anything in the function body)